src/Std/Data/HashSet/Lemmas.lean

```diff
@@ -0,0 +1,274 @@
+  HashMap.Raw.size_le_size_insert h.out
+
+@[simp]
+theorem remove_empty {k : α} {c : Nat} : (empty c : Raw α).remove k = empty c :=
+  ext HashMap.Raw.remove_empty
+
+@[simp]
+theorem remove_emptyc {k : α} : (∅ : Raw α).remove k = ∅ :=
+  ext HashMap.Raw.remove_emptyc
+
+@[simp]
+theorem isEmpty_remove [EquivBEq α] [LawfulHashable α] {k : α} : (m.remove k).isEmpty = (m.isEmpty || (m.size == 1 && m.contains k)) :=
+  HashMap.Raw.isEmpty_remove h.out
+
+@[simp]
+theorem contains_remove [EquivBEq α] [LawfulHashable α] {k a : α} : (m.remove k).contains a = (!(a == k) && m.contains a) :=
+  HashMap.Raw.contains_remove h.out
+
+@[simp]
+theorem mem_remove [EquivBEq α] [LawfulHashable α] {k a : α} : a ∈ m.remove k ↔ (a == k) = false ∧ a ∈ m :=
+  HashMap.Raw.mem_remove h.out
+
+theorem contains_of_contains_remove [EquivBEq α] [LawfulHashable α] {k a : α} : (m.remove k).contains a → m.contains a :=
+  HashMap.Raw.contains_of_contains_remove h.out
+
+theorem mem_of_mem_remove [EquivBEq α] [LawfulHashable α] {k a : α} : a ∈ m.remove k → a ∈ m :=
+  HashMap.Raw.mem_of_mem_remove h.out
+
+theorem size_remove [EquivBEq α] [LawfulHashable α] {k : α} : (m.remove k).size = bif m.contains k then m.size - 1 else m.size :=
+  HashMap.Raw.size_remove h.out
+
+theorem size_remove_le [EquivBEq α] [LawfulHashable α] {k : α} : (m.remove k).size ≤ m.size :=
+  HashMap.Raw.size_remove_le h.out
+
+@[simp]
+theorem fst_containsThenInsert {k : α} : (m.containsThenInsert k).1 = m.insert k :=
+  ext (HashMap.Raw.fst_containsThenInsert h.out)
+
+@[simp]
+theorem snd_containsThenInsert {k : α} : (m.containsThenInsert k).2 = m.contains k :=
+  HashMap.Raw.snd_containsThenInsert h.out
+
+end Raw
+
+section
+
+variable {m : HashSet α}
+
+private theorem ext {m m' : HashSet α} : m.inner = m'.inner → m = m' := by
+  cases m; cases m'; rintro rfl; rfl
+
+@[simp]
+theorem isEmpty_empty {c} : (empty c : HashSet α).isEmpty :=
+  HashMap.isEmpty_empty
+
+@[simp]
+theorem isEmpty_emptyc : (∅ : HashSet α).isEmpty :=
+  HashMap.isEmpty_emptyc
+
+@[simp]
+theorem isEmpty_insert [EquivBEq α] [LawfulHashable α] {a : α} : (m.insert a).isEmpty = false :=
+  HashMap.isEmpty_insert
+
+theorem contains_congr [EquivBEq α] [LawfulHashable α] {a b : α} (hab : a == b) : m.contains a = m.contains b :=
+  HashMap.contains_congr hab
+
+theorem mem_congr [EquivBEq α] [LawfulHashable α] {a b : α} (hab : a == b) : a ∈ m ↔ b ∈ m :=
+  HashMap.mem_congr hab
+
+@[simp] theorem contains_empty {a : α} {c} : (empty c : HashSet α).contains a = false :=
+  HashMap.contains_empty
+
+theorem mem_iff_contains {a : α} : a ∈ m ↔ m.contains a :=
+  HashMap.mem_iff_contains
+
+@[simp] theorem not_mem_empty {a : α} {c} : ¬a ∈ (empty c : HashSet α) :=
+  HashMap.not_mem_empty
+
+@[simp] theorem contains_emptyc {a : α} : (∅ : HashSet α).contains a = false :=
+  HashMap.contains_emptyc
+
+@[simp] theorem not_mem_emptyc {a : α} : ¬a ∈ (∅ : HashSet α) :=
+  HashMap.not_mem_emptyc
```

Comment:
Perhaps shuffle the order here, so each `_empty` / `_emptyc` pair is adjacent?

Reply:
I guess the question is: do you care more about having `empty`/`emptyc` pairs adjacent or having `contains`/`mem` pairs adjacent? I'm happy with either.

Reply:
In either case, `mem_iff_contains` needs to move up one. :-)

For the other question, I'm happy with either now, so no change required.